aten/src/ATen/native/cpu/SpmmReduceKernel.cpp

```diff
@@ -21,6 +21,135 @@ namespace at { namespace native {
 
 namespace {
 
+template <typename scalar_t, typename index_t, ReductionType reduce>
+inline void _update(scalar_t* out_ptr, int64_t row_start, int64_t row_end, const TensorAccessor<index_t, 1>& col_data, const TensorAccessor<scalar_t, 1>& val_data, scalar_t* other_data, int64_t K) {
+  using Vec = vec::Vectorized<scalar_t>;
+  using fVec = vec::Vectorized<float>;
+  constexpr int64_t kVecSize = Vec::size();
+  constexpr int64_t kVLEN = kVecSize * 4;
+  constexpr int64_t CHUNK_SIZE = 16;
+
+  if constexpr (!is_reduced_floating_point_v<scalar_t>) {
```

Comment:
you can use `vec2` to align float and float16/bfloat16 paths into one:

https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/utils.h#L43

when input is float, Vec would be Vectorized<scalar_t>
when input is bfloat16 or float16, Vec would be Vec2